google-cloud-storage/src/main/scala/akka/stream/alpakka/googlecloud/storage/impl/GCStorageStream.scala

```diff
@@ -0,0 +1,640 @@
+
+    createAuthenticatedRequestSource(request)
+  }
+
+  private def createPostSourceRequestSource(
+      contentType: ContentType,
+      data: Source[ByteString, _],
+      uriFactory: GCStorageSettings => Uri,
+      headers: Seq[HttpHeader] = Seq.empty
+  )(implicit mat: ActorMaterializer, attr: Attributes): Source[HttpRequest, NotUsed] = {
+
+    implicit val sys = mat.system
+    val conf = resolveSettings()
+
+    val uri = uriFactory(conf)
+    val request = HttpRequest(HttpMethods.POST)
+      .withUri(uri)
+      .withEntity(HttpEntity(contentType, data))
+      .withHeaders(headers: _*)
+
+    createAuthenticatedRequestSource(request)
+  }
+
+  private def createAuthenticatedRequestSource(request: HttpRequest)(implicit mat: ActorMaterializer,
+                                                                     attr: Attributes): Source[HttpRequest, NotUsed] = {
+    import mat.executionContext
+    implicit val sys = mat.system
+    val conf = resolveSettings()
+    val http = Http()(mat.system)
+    val session: GoogleSession =
+      new GoogleSession(conf.clientEmail,
+                        conf.privateKey,
+                        new GoogleTokenApi(http, TokenApiSettings(conf.tokenUrl, conf.tokenScope)))
+
+    Source.fromFuture(session.getToken().map { accessToken =>
+      request.addCredentials(OAuth2BearerToken(accessToken))
+    })
+  }
+
+  private def getBucketPath(bucket: String) =
+    Path("/b") / bucket
+
+  private def getObjectPath(bucket: String, objectName: String) =
+    getBucketPath(bucket) / "o" / objectName
+
+  private def swap[A](option: Option[Future[A]])(implicit ec: ExecutionContext): Future[Option[A]] =
+    option match {
+      case Some(f) => f.map(Some(_))
+      case None => Future.successful(None)
+    }
+
+  private def responseEntityOptionTo[A](
+      responseOption: Option[ResponseEntity]
+  )(implicit aRead: JsonReader[A], mat: Materializer): Future[Option[A]] = {
+    import mat.executionContext
+    swap(responseOption.map(resp => responseEntityTo(resp)(aRead, mat)))
+  }
+
+  private def responseEntityTo[A](response: ResponseEntity)(implicit aRead: JsonReader[A],
+                                                            mat: Materializer): Future[A] = {
+    import mat.executionContext
+    Unmarshal(response).to[JsValue].map(data => data.convertTo[A])
+  }
+
+  private def entityForSuccessOption(
+      resp: HttpResponse
+  )(implicit mat: Materializer): Future[Option[ResponseEntity]] = {
+
+    import mat.executionContext
+
+    resp match {
+      case HttpResponse(status, _, entity, _) if status.isSuccess() && !status.isRedirection() =>
+        Future.successful(Some(entity))
+      case HttpResponse(StatusCodes.NotFound, _, entity, _) =>
+        entity.discardBytes()
+        Future.successful(None)
+      case HttpResponse(_, _, entity, _) =>
+        Unmarshal(entity).to[String].flatMap { err =>
+          Future.failed(new RuntimeException(err))
```

Comment:
It is usually JSON document that contains information about error and it includes status code, but I think that it could be useful to add status code to be sure.